backend/vouch/trust_algo.py

```diff
@@ -13,98 +13,65 @@
-    """
-    Vouch matrix normalization guarantees three properties:
-        - The sum of normalized vouches given by a voucher equals 1.
-        - Each voucher vouches for pre-trusted users.
-        - Vouchers that explicitly vouch for many barely vouch for pre-trusted
-          users.
-
-    Keyword arguments:
-    vouch_matrix -- A 2 dimensional array of vouch values.
-                    The 1st dimension is the voucher, the 2nd is the vouchee.
-                    vouch_matrix[voucher][vouchee] > 0 if voucher vouched for vouchee.
-    pretrusts -- pretrusts[u] > 0 if u is pretrusted.
-    """
-    normalized_vouch_matrix = np.zeros(vouch_matrix.shape)
-
-    nb_users = len(pretrusts)  # Number of users
-    nb_pretrusted = np.sum(
-        np.array(pretrusts) > 0, axis=0
-    )  # Number of pretrusted users
-
-    # TODO/PERF this could be done with numpy instead of nested for loops (100x speed up)
-    for voucher in range(nb_users):
-        n_vouches_by_voucher = np.sum(np.array(vouch_matrix[voucher]) > 0, axis=0)
-        normalization_constant = (
-            IMPLICIT_PRETRUST_VOUCH * nb_pretrusted + n_vouches_by_voucher
-        )
-        for vouchee in range(nb_users):
-            if pretrusts[vouchee] > 0:
-                normalized_vouch_matrix[voucher][vouchee] += (
-                    IMPLICIT_PRETRUST_VOUCH / normalization_constant
-                )
-            if vouch_matrix[voucher][vouchee] > 0:
-                normalized_vouch_matrix[voucher][vouchee] += 1 / normalization_constant
-    return normalized_vouch_matrix
-
-
-def compute_relative_posttrusts(normalized_vouch_matrix, relative_pretrusts: NDArray):
-    """
-    Return a vector of global trust values per user, given the vouchers in the
-    network and the set of pre-trusted users. This part comes directly from
-    EigenTrust.
+
+def get_weighted_vouch_matrix(vouch_matrix: NDArray) -> NDArray:
     """
-    relative_trusts = relative_pretrusts
-    new_relative_trusts = relative_trusts
-    delta = 10
-    while delta >= APPROXIMATION_ERROR:
-        new_relative_trusts = normalized_vouch_matrix.T.dot(relative_trusts)
+    Returns the weighted_vouch_matrix, derived from the vouch_matrix
+    (composed of explicit Voucher created by users).
 
-        new_relative_trusts = (
-            1 - PRETRUST_BIAS
-        ) * new_relative_trusts + PRETRUST_BIAS * relative_pretrusts
+    The weighted_vouch_matrix integrates implicit vouches, i.e vouches
+    attributed to a sink, such as the sum of weighted vouches given by
+    any voucher is at most one.
+    """
 
-        delta = np.linalg.norm(new_relative_trusts - relative_trusts)
-        relative_trusts = new_relative_trusts
-    return new_relative_trusts
+    n_vouches_by_voucher = vouch_matrix.sum(axis=1)
+    weighted_denominator = n_vouches_by_voucher + SINK_VOUCH
+    weighted_vouch_matrix = vouch_matrix / weighted_denominator[:, np.newaxis]
+    return weighted_vouch_matrix
```

Comment:
The sink and decay parameters seem to have another impact which is how much trust one not pretrusted user gets if vouched by 1, 2, 3, 4 pre-trusted contributors. Currently it seems to be somewhere around `~DECAY/SINK = ~0.1` for getting vouched by a single pre-trusted contributor. I find that this is maybe rather low? 0.25 would be a better target ? Has it been taken into account for the decision of the parameter values?

Reply:
Yes there is a trade-off between not disincentivizing users from vouching (without diluting the scores too fast when adding new vouchees), and allowing non-pretrusted to get significant trust without too many vouchers. 

I suppose the current value `SINK_VOUCH = 10` has been chosen to be very cautious about how trust is assigned. It has been suggested by @glerzing to set a smaller value. Personally I would agree to set `SINK_VOUCH` to 4 or 5.

To be confirmed with Core team.

Reply:
Here was the rationale.
How much trust do you earn from 1 "unique" vouch (where unique means this person only vouched you)? Suppose this voucher is pretrusted and has a trust of T=0.8 and β=0.8 and v = 10 the sink implicit vouching. If my calculation is correct, it gives you T * β * 1/ (1 + v) = 0.8² / (1 + v) = 0.0582
This means that to get fully trusted, you need between 10 and 20 unique vouches. This looks hard to gather, Tournesol is not a social network.

I proposed a lower value for the sink implicit vouching of 3 or 4. Which gives 0.16 or 0.128 per unique vouching. Which is still not a lot actually.
It gives more incentives to vouch for people you really know (e.g. your family). And you don't really loose vouching potential (it's actually the contrary, because less is thrown to the sink), what you loose is the incentive to vouch a lot of people.
But it also makes it more important to give some clue to contributors that the more they vouch, the less weight these vouches carry.

Here is a graph of the **total of trust given by a contributor** n when vouching x times (= 0.8² * x / (x + v)):

![image](https://user-images.githubusercontent.com/47578089/199624192-a0734913-c88c-4195-a7c3-0ef3f23dfe57.png)


And here is a graph of the **value of each individual vouching**, given how many vouches you have done:

![image](https://user-images.githubusercontent.com/47578089/199625000-d3ffb4eb-77a5-4eaa-989e-1e0d363dde3f.png)